File: typescript/kata-machine/src/day1/Queue.ts
class QNode<T> {
	public data: T;
	public next: QNode<T> | null;

	constructor(item: T) {
		this.data = item;
		this.next = null;
	}
}

export default class Queue<T> {
	public length: number;
	public head: QNode<T> | null;
	public tail: QNode<T> | null;

	constructor() {
		this.head = null;
		this.tail = null;
		this.length = 0;
	}

	enqueue(item: T): void {
		const newNode = new QNode(item);

		if (!this.head) {
			this.head = newNode;
			this.tail = newNode;
			this.length++;
			return;
		}

		this.tail!.next = newNode;
		this.tail = newNode;
		this.length++;
	}

	deque(): T | undefined {
		if (!this.head) return undefined;

		const removedNode = this.head;

		this.head = this.head!.next;
		this.length--;

		if (this.length === 0) this.tail = null;
		return removedNode?.data;
	}

	peek(): T | undefined {
		if (!this.head) return undefined;
		return this.head.data;
	}
}

File: C/queue/queue.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   queue.c                                            :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: gusda-si <gusda-si@student.42sp.org.br>    +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2023/07/17 11:08:12 by gusda-si          #+#    #+#             */
/*   Updated: 2023/07/17 12:12:06 by gusda-si         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "queue.h"

extern t_qnode	*ft_queue_create_node(void *data);

void	ft_queue_init(t_queue *queue)
{
	queue->size = 0;
	queue->head = NULL;
	queue->tail = NULL;
}

void	ft_queue_enqueue(t_queue *queue, void *data)
{
	t_qnode	*new_node;

	if (queue->size == 0)
	{
		queue->head = ft_queue_create_node(data);
		queue->tail = queue->head;
		queue->size++;
		return ;
	}
	new_node = ft_queue_create_node(data);
	queue->tail->next = new_node;
	queue->tail = new_node;
	queue->size++;
}

t_qnode	*ft_queue_dequeue(t_queue *queue)
{
	t_qnode	*removed_node;

	if (queue->size == 0)
		return (NULL);
	queue->size--;
	removed_node = queue->head;
	queue->head = queue->head->next;
	if (queue->size == 0)
		queue->tail = NULL;
	removed_node->next = NULL;
	return (removed_node);
}

t_qnode	*ft_queue_peek(t_queue *queue)
{
	if (queue->size == 0)
		return (NULL);
	return (queue->head);
}
